### PR TITLE
ARROW-4245: [Rust] Add Rustdoc header to source files

### PR DESCRIPTION
**Pull request by paulkernfeld**
This commit adds a short description of each module

### CODE REVIEW

**Review by nevi-me — COMMENTED** [review]
Hi @paulkernfeld, thanks for picking this up. I've requested 1 minor change on the header for the csv/mod.rs file.

Please also have a look at the example in array.rs, AppVeyor build failed because of an assertion error. I briefly looked at it, but can't see where the `83` value that it expects comes from.

File: rust/arrow/src/array.rs [hunk 1]
```diff
@@ -15,7 +15,37 @@
 // specific language governing permissions and limitations
 // under the License.
 
-///! Array types
+//! Defines public types representing Apache Arrow arrays. Arrow's specification defines an array as
+//! "a sequence of values with known length all having the same type." For example, the type
+//! `Int16Array` represents an Apache Arrow array of 16-byte integers.
```
Comment (sunchao) [inline rust/arrow/src/array.rs]: `16-byte` -> `16-bit`.

File: rust/arrow/src/array.rs [hunk 2]
```diff
@@ -15,7 +15,37 @@
...
+//!
+//! ```
+//! extern crate arrow;
+//!
+//! use arrow::array::Int16Array;
+//!
+//! // Create a new builder with a capacity of 100
+//! let mut builder = Int16Array::builder(100);
+//!
+//! // Append an individual primitive value
```
Comment (sunchao) [inline rust/arrow/src/array.rs]: nit: `an individual` -> `a single`?

File: rust/arrow/src/buffer.rs [hunk 3]
```diff
@@ -15,6 +15,9 @@
 // specific language governing permissions and limitations
 // under the License.
 
+//! Defines `Buffer`, a contiguous immutable memory region of fixed size aligned at a 64-byte
```
Comment (sunchao) [inline rust/arrow/src/buffer.rs]: Perhaps we should also mention `MutableBuffer` here.

File: rust/arrow/src/lib.rs [hunk 4]
```diff
@@ -15,6 +15,12 @@
 // specific language governing permissions and limitations
 // under the License.
 
+//! A native Rust implementation of [Apache Arrow](https://arrow.apache.org), a cross-language
+//! development platform for in-memory data.
+//!
+//! Currently the project is developed and tested against nightly Rust. To learn more about the
+//! status of Arrow in Rust, see `README.md`.
```
Comment (sunchao) [inline rust/arrow/src/lib.rs]: nit: can `README.md` be a link?

File: rust/arrow/src/array.rs (hunk 1 above)
Comment (paulkernfeld) [reply rust/arrow/src/array.rs]: Yikes, good catch!

File: rust/arrow/src/lib.rs (hunk 4 above)
Comment (paulkernfeld) [reply rust/arrow/src/lib.rs]: I actually don't know how to do this in a portable way. What's the best way to link to the README?

**Comment by paulkernfeld** [comment]
Oops! My test was calling a function that contains some unsafe behavior, so the results of the test were nondeterministic. I have now fixed the test. I think that [ARROW-3776](https://issues.apache.org/jira/browse/ARROW-3776), "Mark methods that do not perform bounds checking as unsafe," could be helpful to others in avoiding a similar pitfall.

I think this PR is ready for another review now.

File: rust/arrow/src/lib.rs (hunk 4 above)
Comment (nevi-me) [reply rust/arrow/src/lib.rs]: Link to the README on master, https://github.com/apache/arrow/blob/master/rust/arrow/README.md

**Review by paddyhoran — COMMENTED** [review]
LGTM, thanks @paulkernfeld.  I had thought about doing this for a while...

**Review by andygrove — APPROVED** [review]
This is looking great. Thanks @paulkernfeld

**Pull request closed without merging**